File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.py
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        freq = {}

        n = len(nums)

        l = 0

        maxSize = 0

        for r in range(0 , n):
            freq[nums[r]] = freq.get(nums[r] , 0) + 1

            while freq.get(nums[r]) > k:
                freq[nums[l]] = freq.get(nums[l] , 0) - 1;

                if freq[nums[l]] == 0:
                    del freq[nums[l]]

                l += 1

            maxSize = max(maxSize , r - l + 1)

        return maxSize
        

File: 2958-length-of-longest-subarray-with-at-most-k-frequency/2958-length-of-longest-subarray-with-at-most-k-frequency.java
class Solution {
    public int maxSubarrayLength(int[] nums, int k) {
        HashMap<Integer , Integer> freq = new HashMap<>();

        int n = nums.length;

        int maxSize = 0;

        int l = 0;

        for(int r = 0 ; r < n ; r++){
            freq.put(nums[r] , freq.getOrDefault(nums[r] , 0) + 1);

            while(freq.get(nums[r]) > k){
                freq.put(nums[l] , freq.get(nums[l]) - 1);
                l++;
            }

            maxSize = Math.max(maxSize , r - l + 1);
        }

        return maxSize;
    }
}